Meme/Meme/MemeTemplatesViewController.swift

```diff
@@ -38,6 +54,33 @@ class MemeTemplatesViewController: UIViewController {
     }
 }
 
+extension MemeTemplatesViewController : UITableViewDelegate, UITableViewDataSource {
+
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return self.templates.count
+    }
+
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = self.tableView.dequeueReusableCell(withIdentifier: cellID) as! MemeTemplatesTableViewCell
+        cell.memeImageView.image = nil
+
+        if let imageURL = self.templates[indexPath.row].imageURL {
+            if let cachedImage = self.imageCache[imageURL.absoluteString] {
+                cell.memeImageView.image = cachedImage
+            } else {
+                cell.memeImageView.load(url: imageURL) { image in
+                    self.imageCache[imageURL.absoluteString] = image
+                    if let cellToUpdate = self.tableView.cellForRow(at: indexPath) as? MemeTemplatesTableViewCell {
```

Comment:
should probably get the cell from the visible cells list rather than force reloading the cell

Reply:
<img width="966" alt="Screen Shot 2019-08-18 at 12 07 40 PM" src="https://user-images.githubusercontent.com/5643468/63227082-cc72ef00-c1b0-11e9-8f55-a712a04ffc9a.png">
<img width="896" alt="Screen Shot 2019-08-18 at 12 07 22 PM" src="https://user-images.githubusercontent.com/5643468/63227083-d1d03980-c1b0-11e9-85bd-c82436841f21.png">

Reply:
Was `self.tableview.visibleCells[indexPath.row]` what you were suggesting or was it something else? I'm getting of by 1 errors. Currently it crashes whenever I scroll, if I update the code with that. Or if I change it to `indexPath.row-1` then it crashes on startup.

Reply:
I was thinking more like 
```swift
if (tableView.visibleCells.contains(cell) { update ImageView }
```

Reply:
<img width="559" alt="Screen Shot 2019-08-18 at 3 12 13 PM" src="https://user-images.githubusercontent.com/5643468/63229625-84fc5b00-c1d0-11e9-85fc-0d66db865c58.png">
<img width="559" alt="Screen Shot 2019-08-18 at 3 12 17 PM" src="https://user-images.githubusercontent.com/5643468/63229627-84fc5b00-c1d0-11e9-99f0-0268b379feca.png">

This is what I get when I scroll if I use `tableview.visibleCells.contains(cell) {...}`

Reply:
hmm let me take a look. For reference, scroll really fast and you'll see the issue we're trying to prevent. the cells will flash repeatedly as images are retrieved

Reply:
this seems to work for me
```swift
cell.memeImageView.load(url: imageURL) { [weak cell, weak tableView] image in
                guard let visiblePaths = tableView?.indexPathsForVisibleRows,
                    let cell = cell,
                    visiblePaths.contains(indexPath)
                    else { return }
                cell.memeImageView.image = image
                cell.layoutIfNeeded()
            }
```

Reply:
Gotcha, yeah that worked for me too. Do we need to weakify tableView? The cell is not holding a reference to it... Also, could you explain the difference between `setNeedsLayout` and `layoutIfNeeded` and when you'd pick one over the other?

Reply:
I weakified the tableview because the tableview references the cells
TBLV -> Cell -> memeImageView -> loadClosure -> tableview/cell

Reply:
`setNeedsLayout` is an async request to update layout on the next render cycle. `layoutIfNeeded` is a sync request to _immediately_ rerender. I used `layoutIfNeeded` here because the imageView sometimes wasn't laying out when using `setNeedsLayout`. More conventionally I'd use layoutIfNeeded to update constraints in an animation block, but it worked in this case also. 

More on this:
https://medium.com/@varunrathi_78758/ios-setneedslayout-vs-layoutifneeded-ceb1c8335abe

Reply:
Thanks for the explanation!